01-dp: Задача о разбиении строки (1)

riqkum opened the pull request.

riqkum committed:
01 Text segmentation problem (1)

riqkum committed:
02 Topological sort

vseloved commented on tasks/max_beznos/02-graph-dp.lisp:
```diff
@@ -0,0 +1,35 @@
+                 (visit v))))
+      (loop
+         :for h :being :each hash-key :in visited
+         :collect h))))
```
ты здесь расчитываешь на то, что хеш-таблица отсортированна в порядке добавления ключей в нее?

vseloved commented on tasks/max_beznos/01.lisp:
```diff
@@ -0,0 +1,73 @@
+                           (let* ((hint (find-if (lambda (h) (char= (first h) (char text 0)))
+                                                 hint-list))
+                                  (entry-index (match-entry dictionary-array text
+                                                            :start (if start start (second hint))
```
`(or start (second hint))`

riqkum committed:
Corrections after review (01, 02)

- topological sort: maintain proper precedence while traversing in DFS

riqkum committed:
01 Text segmentation: support selecting ngrams with maximal rate

vseloved commented on tasks/max_beznos/02-graph-dp.lisp:
```diff
@@ -0,0 +1,37 @@
+                   :collect (child edge))))
+    (let ((visited  (make-hash-table))
+          (traverse (list))
+          (vertices (remove-duplicates
```
не очень эффективно, т.к. `remove-duplicates` работает за супердинейное время, кажется. Поэтому я даже использую отдельную утилиту (как правило): https://github.com/vseloved/cl-nlp/blob/master/src/util/misc.lisp#L41

vseloved commented on tasks/max_beznos/02-graph-dp.lisp:
```diff
@@ -0,0 +1,37 @@
+                          (visit n)))
+                 (push vertex traverse)
+                 (setf (gethash vertex visited) 'visited)))
+        (loop
```
тут тоже, в идеале, не мешало бы придумать сублинейный вариант выбора следующей вершины. Я это сделал через множество unvisited, из которого беру всегда первую случайную вершину

vseloved commented on tasks/max_beznos/01.lisp:
```diff
@@ -0,0 +1,106 @@
+                           :until (or (eql record '*eof*)
+                                      (and head (= count head)))
+                           :collect (parse-dictionary-record record))))))
+      (make-array (list (length dict-list) 4)
```
проблема тут, как мы говорили, что поиск по рейнджу будет существенно медленней поиска по хеш-таблице

vseloved commented on tasks/max_beznos/01.lisp:
```diff
@@ -0,0 +1,106 @@
+               (setf range-first-char (char entry 0)))))
+       finally (return (reverse result)))))
+
+(defun match-entry (dictionary-array-with-rates text &key start end (max-rate nil))
```
идея с `max-rate` тоже не до конца ясна

riqkum committed:
01 Text segmentation: algorithm is completely changed

- store dictionary in a hash table
- precise edges encoding using 1d flat array
- edge relaxation using Bellman-Ford algorithm

vseloved commented:
Интересное компрессированное представление графа, можешь его описать текстом (и/или картинками), чтобы можно было обсудить его свойства и скинуть другим?
